### DISCUSSION TITLE
Another trigger

### DISCUSSION THREAD

**Discussion by Callahan78** [ybizeul/StreamDeckWS#25] [Q&A]
I am pretty new to Node Red and I am getting nowhere even though I have tried many times.
I have reproduced the example program and it works.
Now I want to customize it for my needs and replace the web socket trigger with an "ioBroker in" (trigger on state change). Unfortunately the triggering does not work. Could someone tell me what is still missing here?

![Unbenannt](https://user-images.githubusercontent.com/88541621/170846771-80274dbf-5cf9-4a7d-880d-f0566f061898.JPG)

**Comment by ybizeul** [comment]
Sorry for the delayed answer.

Best practice is to use the new `sd-input` and `sd-output` node-red nodes from [node-red-contrib-streamdeck-ws](https://flows.nodered.org/node/node-red-contrib-streamdeck-ws) so the flow can manage internal IDs of the different button when they load, and cache it for further use.

If you don't use those nodes, you will have to be careful injecting the right parameters to the websocket message when returning. Those parameters, in particular the `context` id come from the websocket, which explains you don't have it hen using another trigger.

One way to fix this is to catch all the buttons loading when StreamDeck initializes and cache in a flow variable all the `context` and `settings/id` sent alongside the event. Again, that's when the StreamDeck initializes.

Later on, when you need to change a button state, you fetch in that cache what is the `context` for this button and make sure you add it to the message sent to the StreamDeck.

This is a bit complicated, that's why I created the [node-red-contrib-streamdeck-ws](https://flows.nodered.org/node/node-red-contrib-streamdeck-ws) module that you can use like this :

![image](https://user-images.githubusercontent.com/182035/171944319-4b0d5968-5d7d-4ccd-a7c0-f63653d28b64.png)

Basically you send all messages coming from StreamDeck to `sd-input` and then use `sd-output` to sent new states to StreamDeck :

![image](https://user-images.githubusercontent.com/182035/171944887-de904cb9-869b-4ae1-9582-3db1612f75f2.png)

**Comment by Callahan78** [comment]
Thank you very much for the further hints. I will now see how far I can get with it.

Maybe I'll shortly explain my goal:

Simple example: control a lamp.

Goal 1:
Switching the lamp via stream deck. Here I would probably start with sd-input and then have to change the object state (on/off) in ioBroker.

Goal 2:
I want the stream deck to display the status of the lamp (on/off image in the stream deck). When the status of the lamp changes (whether by stream deck or by other trigger), the right picture should be displayed in stream deck. This is the need I described in my input post.

A later variant of goal 2 is not to change an image in the stream deck, but display a concrete value from an ioBroker object (e.g. a temperature).

If I had a working template for all this, I could create a lot of cool actions for it :-)

**Comment by ybizeul** [comment]
I don't use ioBroker, but I guess HomeAssistant module is similar. That's basically the example above in my answer, then :

change: 2 rules is like this :
![image](https://user-images.githubusercontent.com/182035/171992765-d0a683e2-2484-4fdf-a7c9-79bcdda9644f.png)

And `sd-output` __SttreamDeck__ is :
![image](https://user-images.githubusercontent.com/182035/171992788-f9aabe2c-3921-4fd7-904d-dac8ebfec9eb.png)

**Comment by Callahan78** [comment]
I have now installed node-red-contrib-streamdeck-ws. But to be honest, haven't quite understood how to use that to achieve my goal.
I can't quite interpret your examples either. So I think I need to get more basics in node red to get further here.

**Comment by ybizeul** [comment]
It is a little bit convoluted indeed. the `sd-input` node "does" nothing in itself, it's just getting request and caches them in the follow for further use by `sd-output`. Not sure it makes sense !

**Comment by Callahan78** [comment]
It surely makes sense, but I don't quite understand it yet ;-)
If I understood it correctly, sd-input/sd-output is only for analysis of the StreamDeck and not for later use in the production script. (?)

**Comment by ybizeul** [comment]
No they are for production use. Bottomline is : you put a `sd-input` right after the the websocket node, and use `sd-output` to update your StreamDeck.
